tests/register/register.go

```diff
@@ -31,6 +38,97 @@ func init() {
 	Tests = make(map[TestType][]types.Test)
 }
 
-func Register(tType TestType, t types.Test) {
+func register(tType TestType, t types.Test) {
 	Tests[tType] = append(Tests[tType], t)
 }
+
+// Registers t multiple times with different, compatible config versions
+func Register(tType TestType, t types.Test) {
+	// todo: how to handle errors? it errors on Empty Userdata, Preemption with default config, Preemption with base, default config, Preemption with no config but these test still need to run
+	var config typesInternal.Config
+	err := json.Unmarshal([]byte(t.Config), &config)
+	version, semverErr := semver.NewVersion(config.Ignition.Version)
+
+	register(tType, t)
+	if err == nil && semverErr == nil && version.LessThan(maxTypes.MaxVersion) { // todo: how to identify max of different major versions?
+		versionStr := strconv.FormatInt(version.Major, 10) + "." + strconv.FormatInt(version.Minor, 10) + "." + strconv.FormatInt(version.Patch, 10)
+
+		for *version != maxTypes.MaxVersion {
+			test := deepCopy(t)
+
+			version.Minor++
+			updatedVersionStr := strconv.FormatInt(version.Major, 10) + "." + strconv.FormatInt(version.Minor, 10) + "." + strconv.FormatInt(version.Patch, 10)
+			if version.Minor == 3 {
+				*version = maxTypes.MaxVersion
+				updatedVersionStr += "-" + string(version.PreRelease) + version.Metadata
+			}
+
+			test.Config = strings.Replace(test.Config, versionStr, updatedVersionStr, 1)
+			test.Name += " " + updatedVersionStr
+			register(tType, test)
+		}
+	}
+}
+
+// Deep copy Test struct fields In, Out, MntDevices, OEMLookasideFiles, SystemDirFiles
+// so each BB test with identical Test structs have their own independent Test copies
+func deepCopy(t types.Test) types.Test {
+	In_diskArr := make([]types.Disk, len(t.In))
+	copy(In_diskArr, t.In)
+	t.In = deepCopyPartitions(In_diskArr)
+
+	Out_diskArr := make([]types.Disk, len(t.Out))
+	copy(Out_diskArr, t.Out)
+	t.Out = deepCopyPartitions(Out_diskArr)
+
+	mntdevice := make([]types.MntDevice, len(t.MntDevices))
+	copy(mntdevice, t.MntDevices)
+	t.MntDevices = mntdevice
```

Comment:
Is it possible to make a function using generic variables like in other languages? There wasn't anything super straight forward online, but I'll dig into it more over the next day or so.